dkulib/dku_io_utils/chunked_processing.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+"""Module with read/write utility for chunked processing based on the Dataiku API"""
+
+import logging
+import math
+from time import perf_counter
+from typing import Callable
+
+from tqdm.auto import tqdm as tqdm_auto
+import dataiku
+
+
+def count_records(dataset: dataiku.Dataset) -> int:
+    """Count the number of records of a dataset using the Dataiku dataset metrics API
+
+    Args:
+        dataset: dataiku.Dataset instance
+
+    Returns:
+        Number of records
+
+    """
+    metric_id = "records:COUNT_RECORDS"
+    partitions = dataset.read_partitions
+    client = dataiku.api_client()
+    project = client.get_project(dataset.project_key)
+    record_count = 0
+    logging.info(f"Counting records of dataset: {dataset.name}...")
+    if partitions is None or len(partitions) == 0:
+        project.get_dataset(dataset.short_name).compute_metrics(metric_ids=[metric_id])
+        metric = dataset.get_last_metric_values()
+        record_count = dataiku.ComputedMetrics.get_value_from_data(metric.get_global_data(metric_id=metric_id))
+        logging.info(f"Dataset {dataset.name} contains {record_count:d} records and is not partitioned")
+    else:
+        for partition in partitions:
+            project.get_dataset(dataset.short_name).compute_metrics(partition=partition, metric_ids=[metric_id])
+            metric = dataset.get_last_metric_values()
+            record_count += dataiku.ComputedMetrics.get_value_from_data(
+                metric.get_partition_data(partition=partition, metric_id=metric_id)
+            )
+        logging.info(f"Dataset {dataset.name} contains {record_count:d} records in partition(s) {partitions}")
+    return record_count
+
+
+def process_dataset_chunks(
+    input_dataset: dataiku.Dataset, output_dataset: dataiku.Dataset, func: Callable, chunksize: float = 1000, **kwargs
+) -> None:
+    """Read a dataset by chunks, process each dataframe chunk with a function and write back to another dataset.
+
+    Pass keyword arguments to the function, adds a tqdm progress bar and generic logging.
+    Directly write chunks to the output_dataset, so that only one chunk needs to be processed in-memory at a time.
```

Comment:
Perhaps a stupid question - but what's the key differentiation from DataFrameParallelizer in [dkulib/parallelizer](https://github.com/dataiku/dss-plugin-dkulib/blob/feature/parallelizer/dkulib/parallelizer/parallelizer.py)?

I see:
- One chunk in memory at a time - Might also be possible to do in DataFrameParallelizer, no?
- Parallelization -- can be off if workers is just 1
- Feeding in a dataiku.Dataset vs pd.DataFrame - could be changed in one or the other

Wouldn't it be easier for users to have just one go-to wrapper for applying functions to dataiku datasets?

Reply:
Good remark! While there are some similarities, there are actually several differences:
1. DataFrameParallelizer works on pandas.DataFrame, while process_dataset_chunks works on dataiku.Dataset. 
2. DataFrameParallelizer can handle errors, so is well suited for failure-prone API calls. In contrast, process_dataset_chunks will fail upon any error from the underlying function.
3. DataFrameParallelizer enforces a set output schema, designed for API calls with responses and errors. Meanwhile, process_dataset_chunks lets the user set the schema of the output DataFrame.
   - For the record, this is due to the fact both approaches were designed for different needs: DataFrameParallelizer was designed for API-related plugins, whereas process_dataset_chunks was designed for pure Python NLP processing.
4. DataFrameParallelizer can parallelize (row-by-row or by batches of rows) whereas process_dataset_chunks is working sequentially

I chose to write both modules separately mainly because:
(i) I find value in the separation of concerns between internal Dataiku API stuff and pure Python pandas stuff. In particular, we don't currently have a way to test dkulib modules if they rely on the internal Dataiku API.
(ii) End users can freely choose to use one or the other, or both at the same time:
```
process_dataset_chunks(
    input_dataset=dataiku.Dataset("input"),
    output_dataset=dataiku.Dataset("output"),
    func=parallelizer.run
)
```

Now, I agree there would be long-term value in merging the two concepts in a single module. I would definitely log that as an idea for future improvement. Not necessarily a must-have for right now (if it's not broken, ...) but definitely something to keep in mind.

What do you think?

Reply:
Good with me!
Thanks for the detailed explanation 😊